compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -48,20 +49,34 @@ bool isValidSchedule(const GPUMatmulShapeType &problem,
   bool isValidN = (alignedNSize % (schedule.nSize * schedule.nTileCount *
                                    schedule.nWarpCount)) == 0;
   bool isValidK = (alignedKSize % (schedule.kSize * schedule.kTileCount)) == 0;
-  return isValidN && isValidM && isValidK;
+
+  // Constraint to ensure wgTileSize is distributable by wgSize.
+  // such that we can distribute to it's corresponding vector.transfer_read.
+  // TODO: Add constraint on M-dim and K-dim if it pops up in another
+  // model/example. Currently we are being conservative to preserve performance
+  // as much as we can.
+  const int64_t kMaxVectorLoadBitWidth = 128;
+  int64_t elemsPerThread =
+      kMaxVectorLoadBitWidth / problem.bType.getIntOrFloatBitWidth();
+  int64_t wgThreads = schedule.mWarpCount * schedule.nWarpCount * subgroupSize;
+  int64_t nTileSize =
+      schedule.nSize * schedule.nTileCount * schedule.nWarpCount;
+  bool isDistributableN = (nTileSize / elemsPerThread) % wgThreads == 0 ||
+                          wgThreads % (nTileSize / elemsPerThread) == 0;
```

Comment:
you commented that this will work for `matmul_transpose_b` but not for `matmul` because it depends on which dimension is inner most. Can we add identifiers for which dimension is inner most to `GPUMatmulShapeType` to inform this heuristic about when to check for this?

Reply:
Yeah, we can do that let me think of a nice way to shuttle this data around. :)

Still no guarantees(though unlikely) that somewhere down the line this information may change, but defo better than no heuristics haha.